systems/controllers/constrained_lqr_controller.cc

```diff
@@ -1,103 +1,95 @@
-  DRAKE_DEMAND(Q.rows() == num_positions_ + num_velocities_);
+  DRAKE_DEMAND(Q.rows() == plant_.num_positions() + plant_.num_velocities());
   DRAKE_DEMAND(Q.rows() == Q.cols());
-  DRAKE_DEMAND(R.rows() == num_efforts_);
+  DRAKE_DEMAND(R.rows() == plant_.num_actuators());
   DRAKE_DEMAND(R.rows() == R.cols());
 
-  // Creating the full state vector (Velocities are zero as it is a fixed point)
-  VectorXd x(num_states_);
-  VectorXd v = VectorXd::Zero(num_velocities_);
-  x << q, v;
-  desired_state_ = x;
-
-  // Computing the full Jacobian (Position and contact)
-  MatrixXd J_tree_qdot, J_tree_v, J_contact_qdot, J_contact_v;
-  KinematicsCache<double> k_cache = tree_.doKinematics(q);
+  auto J_active_v = evaluators_.EvalActiveJacobian(context);
 
-  // Two separate Jacobians -- in terms of qdot and v are computed as the number
-  // of positions and velocities may be different (Quaternion base model)
-
-  // Position Jacobian
-  J_tree_qdot = tree_.positionConstraintsJacobian(k_cache, true);
-  J_tree_v = tree_.positionConstraintsJacobian(k_cache, false);
+  // convert to w.r.t. qdot one column at a time
+  MatrixX<AutoDiffXd> J_active_qdot(J_active_v.rows(), plant_.num_positions());
+  for (int i = 0; i < plant_.num_positions(); i++) {
+    AutoDiffVecXd v_i(plant_.num_velocities());
+    AutoDiffVecXd qdot = AutoDiffVecXd::Zero(plant_.num_positions());
+    qdot(i) = 1;
+    plant_.MapQDotToVelocity(context, qdot, &v_i);
+    J_active_qdot.col(i) = J_active_v * v_i;
+  }
 
-  // Contact Jacobian (If contact information is provided)
-  if (contact_info_.num_contacts > 0) {
-    J_contact_qdot = autoDiffToValueMatrix(
-        contact_toolkit_->CalcContactJacobian(initializeAutoDiff(x), true));
-    J_contact_v = autoDiffToValueMatrix(
-        contact_toolkit_->CalcContactJacobian(initializeAutoDiff(x), false));
+  // Add quaternion constraints to F
+  // Constraints come from the kinematic fact that ||q|| = 1, linearized is
+  // dot(q_0, delta_q) = 0
+  // Note that there is no corresponding constraint on angular velocity, which
+  // is already 3-dimensional (not 4).
+  int num_quat = 0;
+  std::vector<int> quat_start;
+  auto bodies = plant_.GetFloatingBaseBodies();
+  for (auto body : bodies) {
+    if (plant_.get_body(body).has_quaternion_dofs()) {
+      num_quat++;
+      quat_start.push_back(plant_.get_body(body).floating_positions_start());
+    }
   }
 
-  // Stacking them up
-  MatrixXd J_qdot(J_tree_qdot.rows() + J_contact_qdot.rows(),
-                  J_tree_qdot.cols());
-  MatrixXd J_v(J_tree_v.rows() + J_contact_v.rows(), J_tree_v.cols());
-  J_qdot << J_tree_qdot, J_contact_qdot;
-  J_v << J_tree_v, J_contact_v;
+  MatrixXd F_quat =
+      MatrixXd::Zero(num_quat, J_active_qdot.cols() + J_active_v.cols());
+  for (int i = 0; i < num_quat; i++) {
+    F_quat.row(i).segment(quat_start.at(i), 4) = autoDiffToValueMatrix(
+        plant_.GetPositions(context).segment(quat_start.at(i),4));
+  }
 
   // Computing F
```

Comment:
Maybe it's helpful to have something like this:
```
// F = [J_active_qdot, 0         ]
//     [0            , J_active_v]
//     [          F_quat         ]
```

Also, from equation (23) of the paper,  F seems to be
```
// F = [J_active_qdot, 0              ]
//     [d/dt J_active_qdot, J_active_v]
//     [            F_quat            ]
```
so the implementation here is missing a `d/dt J`?

Reply:
I guess `d/dt J_active_qdot=0` because the nominal trajectory is stationary.

Reply:
`d/dt J = 0` is constant for the time-invariant version.

Added that helpful comment.